SYCL/ESIMD/api/functional/ctors/ctor_move.cpp

```diff
@@ -0,0 +1,232 @@
+// XRUN: %clangxx -fsycl %s -fsycl-device-code-split=per_kernel -o %t.out
+// XRUN: %GPU_RUN_PLACEHOLDER %t.out
+// RUN: false
+// XFAIL: *
+// TODO Remove once the simd vector provides move constructor
+//
+// Test for esimd move constructor
+// The test creates source simd instance with reference data and invokes move
+// constructor in different C++ contexts to create a new simd instance from the
+// source simd instance. It is expected for a new simd instance to store
+// bitwise same data as the one passed as the source simd constructor.
+
+// The test proxy is used to verify the move constructor was called actually.
+#define __ESIMD_ENABLE_TEST_PROXY
+
+#include "common.hpp"
+#include <algorithm>
+#include <cassert>
+
+using namespace sycl::ext::intel::experimental::esimd;
+using namespace esimd_test::api::functional::ctors;
+using namespace esimd_test::api::functional;
+
+// The core test functionality.
+// Runs a TestCaseT, specific for each C++ context, for a simd<DataT,NumElems>
+// instance
+template <typename DataT, int NumElems, typename TestCaseT> class run_test {
+  using KernelName = Kernel<DataT, NumElems, TestCaseT>;
+
+public:
+  bool operator()(sycl::queue &queue, const std::string &data_type) {
+    bool passed = true;
+    bool was_moved = false;
+
+    const shared_allocator<DataT> data_allocator(queue);
+    const shared_allocator<int> flags_allocator(queue);
+    const auto reference = generate_ref_data<DataT, NumElems>();
+
+    shared_vector<DataT> input(reference.cbegin(), reference.cend(),
+                               data_allocator);
+    shared_vector<DataT> result(reference.size(), data_allocator);
+
+    // We need a special handling for case of simd<T,1>, as we need to check
+    // more than a single data value; therefore we need to loop over the
+    // reference data to run test
+    if constexpr (NumElems == 1) {
+      const auto n_checks = input.size();
+      const sycl::range<1> range(n_checks);
+
+      // We need a separate flag per each check to have a parallel_for possible,
+      // because any concurrent access to the same memory location is UB, even
+      // in case we are updating variable to the same value in multiple threads
+      shared_vector<int> flags_storage(n_checks, flags_allocator);
+
+      // run check for each of the reference elements
+      queue.submit([&](sycl::handler &cgh) {
+        const DataT *const ptr_in = input.data();
+        const auto ptr_out = result.data();
+        const auto ptr_flags = flags_storage.data();
+
+        cgh.parallel_for<KernelName>(
+            range, [=](sycl::id<1> id) SYCL_ESIMD_KERNEL {
+              const auto work_item_index = id[0];
+              const DataT *const in = ptr_in + work_item_index;
+              const auto out = ptr_out + work_item_index;
+              const auto was_moved_flag = ptr_flags + work_item_index;
+              *was_moved_flag = run_check(in, out);
+            });
+      });
+      queue.wait_and_throw();
```

Comment:
please leave a comment that described what happened in this code

Reply:
Right, please take a look onto 1e54ab1e38961f2b66bd4b6e38acbb254f3856ed